backend/setup.js

```diff
@@ -5,111 +5,145 @@ const logger              = require('./logger').setup;
-					logger.error('Could not write JWT key pair to config file: ' + filename);
-					reject(err);
-				} else {
-					logger.info('Wrote JWT key pair to config file: ' + filename);
+		config_data.jwt = {
+			key: key.exportKey('private').toString(),
+			pub: key.exportKey('public').toString()
+		};
 
-					logger.warn('Restarting interface to apply new configuration');
-					process.exit(0);
-				}
-			});
+		// Write config
+		fs.writeFile(filename, JSON.stringify(config_data, null, 2), (err) => {
+			if (err) {
+				logger.error('Could not write JWT key pair to config file: ' + filename);
+				reject(err);
+			} else {
+				logger.info('Wrote JWT key pair to config file: ' + filename);
 
-		} else {
-			// JWT key pair exists
-			if (debug_mode) {
-				logger.debug('JWT Keypair already exists');
+				logger.warn('Restarting interface to apply new configuration');
+				process.exit(0);
 			}
+		});
 
-			resolve();
+	} else {
+		// JWT key pair exists
+		if (debug_mode) {
+			logger.debug('JWT Keypair already exists');
 		}
-	})
-		.then(() => {
+
+		resolve();
+	}
+}
+
+function setupDefaultUser() {
+	(userModel
+		.query()
+		.select(userModel.raw('COUNT(`id`) as `count`'))
+		.where('is_deleted', 0)
+		.first()
+	).then( (row) => {
+		if (!row.count) {
+			// Create a new user and set password
+			logger.info('Creating a new user: admin@example.com with password: changeme');
+
+			let data = {
+				is_deleted: 0,
+				email:      'admin@example.com',
+				name:       'Administrator',
+				nickname:   'Admin',
+				avatar:     '',
+				roles:      ['admin']
+			};
+
 			return userModel
 				.query()
-				.select(userModel.raw('COUNT(`id`) as `count`'))
-				.where('is_deleted', 0)
-				.first();
-		})
-		.then((row) => {
-			if (!row.count) {
-				// Create a new user and set password
-				logger.info('Creating a new user: admin@example.com with password: changeme');
-
-				let data = {
-					is_deleted: 0,
-					email:      'admin@example.com',
-					name:       'Administrator',
-					nickname:   'Admin',
-					avatar:     '',
-					roles:      ['admin']
-				};
+				.insertAndFetch(data)
+				.then( (user) => {
+					return authModel
+						.query()
+						.insert({
+							user_id: user.id,
+							type:    'password',
+							secret:  'changeme',
+							meta:    {}
+						})
+						.then(() => {
+							return userPermissionModel
+								.query()
+								.insert({
+									user_id:           user.id,
+									visibility:        'all',
+									proxy_hosts:       'manage',
+									redirection_hosts: 'manage',
+									dead_hosts:        'manage',
+									streams:           'manage',
+									access_lists:      'manage',
+									certificates:      'manage'
+								});
+						});
+				})
+				.then(() => {
+					logger.info('Initial admin setup completed');
+				});
+		} else if (debug_mode) {
+			logger.debug('Admin user setup not required');
+		}
+	});
+}
 
-				return userModel
+function setupDefaultSettings() {
```

Comment:
Not quite sure why you took this from a Migration and put it here, you're not doing that to any other migrations...

Reply:
This is the only migration where the concept was broken. I didn't needed to do it in any other migration bcs all of the other migrations just modify the schemas, and not trying to insert things to it. (BTW; we insert users in the setup but we insert settings in a migration?)

As I understand when you run the migrations it connects to the db. And when you run the `settingModel.query().insert()`  it tries to connect to the db again (while we still migrate). Which in multithreaded env (like pg or mysql) not a big deal, you can have multiple connections/connection pools to the same db. But in sqlite you hava a lock on the file, so while one connection is attached to the db, the other connection "can't" connect to it.

If you are interested in it deeper I referenced an issue in the commit message.

Reply:
Ah ok I get your point about breaking the concept. All good.

To clarify, "migrations" should be changes that occur before the App really runs. It's used on the Knex layer and should not perform any logic on data and assumes the database is in a certain state prior to execution.

Setup users is part of the app because it has to check if there is an admin user and only when not found, it creates one. As this happens on startup each time, a migration is not the right place for that.

Reply:
I understand this, BUT if injecting users is a runtime check, inserting settings why not? Also if inserting settings is essential to start the app, I think the default user is essential to start the app too. It breaks the logic if one or the other handled differently. I think it's mostly a theoretical question if inserting default data is part of the db migration or not, and most of the JVM world would place them to the migrations (including myself), but here the code already introduced both ways, and I simply chose the comfy side.